lightning/src/chain/channelmonitor.rs

```diff
@@ -1763,9 +1769,23 @@ impl<ChanSigner: ChannelKeys> ChannelMonitor<ChanSigner> {
 		// Determine new outputs to watch by comparing against previously known outputs to watch,
 		// updating the latter in the process.
 		watch_outputs.retain(|&(ref txid, ref txouts)| {
-			let output_scripts = txouts.iter().map(|o| o.script_pubkey.clone()).collect();
-			self.outputs_to_watch.insert(txid.clone(), output_scripts).is_none()
+			let idx_and_scripts = txouts.iter().map(|o| (o.0, o.1.script_pubkey.clone())).collect();
+			self.outputs_to_watch.insert(txid.clone(), idx_and_scripts).is_none()
 		});
+		#[cfg(test)]
+		{
+		        // If we see a transaction for which we registered outputs previously,
+			// make sure the registered scriptpubkey at the expected index match
+			// the actual transaction output one. We failed this case before #653.
+			for tx in &txn_matched {
+				if let Some(outputs) = self.get_outputs_to_watch().get(&tx.txid()) {
+				        for (idx, idx_and_script) in outputs.iter().enumerate() {
+				                assert!(idx < tx.output.len());
+				                assert_eq!(tx.output[idx].script_pubkey, idx_and_script.1);
```

Comment:
Shouldn't the idx here and the next line up be idx_and_script.0 as well? Then we could drop the enumerate.

Reply:
Correct, if the idx_and_script.0 is wrong it should lead to a un-matching scriptpubkey output on the previously-registered transaction.